server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java

```diff
@@ -45,4 +66,202 @@ public void testZen1NodesJoiningZen2Cluster() {
         internalCluster().startNodes(randomIntBetween(1, 3), ZEN1_SETTINGS);
         createIndex("test");
     }
+
+    public void testMixedClusterFormation() throws Exception {
+        final int zen1NodeCount = randomIntBetween(1, 3);
+        final int zen2NodeCount = randomIntBetween(1, 3);
+        logger.info("starting cluster of [{}] Zen1 nodes and [{}] Zen2 nodes", zen1NodeCount, zen2NodeCount);
+        final List<String> nodes = internalCluster().startNodes(IntStream.range(0, zen1NodeCount + zen2NodeCount)
+            .mapToObj(i -> i < zen1NodeCount ? ZEN1_SETTINGS : ZEN2_SETTINGS).toArray(Settings[]::new));
+
+        createIndex("test",
+            Settings.builder()
+                .put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), TimeValue.ZERO) // assign shards
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, zen1NodeCount + zen2NodeCount) // causes rebalancing
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
+                .build());
+        ensureGreen("test");
+
+        for (final String node : nodes) {
+            final boolean addedVotingTombstone;
+            if (zen1NodeCount == 1 && zen2NodeCount > 1 && zen2NodeCount < 3) {
```

Comment:
these conditions are too complicated for me to understand. Can you add some comment explaining why they're set up that way?

Reply:
Yes, clarified them a bit.